core/services/blockhashstore/feeder.go

```diff
@@ -0,0 +1,164 @@
+}
+
+// Feeder checks recent VRF coordinator events and stores any blockhashes for blocks within
+// waitBlocks and lookbackBlocks that have unfulfilled requests.
+type Feeder struct {
+	logger         logger.Logger
+	coordinator    Coordinator
+	bhs            BHS
+	waitBlocks     int
+	lookbackBlocks int
+	latestBlock    func(ctx context.Context) (uint64, error)
+
+	stored       map[uint64]struct{}
+	lastRunBlock uint64
+}
+
+// Run the feeder.
+func (f *Feeder) Run(ctx context.Context) {
+	latestBlock, err := f.latestBlock(ctx)
+	if err != nil {
+		f.logger.Errorw("Failed to fetch current block number", "error", err)
+		return
+	}
+
+	var (
+		fromBlock        = int(latestBlock) - f.lookbackBlocks
+		toBlock          = int(latestBlock) - f.waitBlocks
+		blockToRequests  = make(map[uint64]map[string]struct{})
+		requestIDToBlock = make(map[string]uint64)
+	)
+	if fromBlock < 0 {
+		fromBlock = 0
+	}
+	if toBlock < 0 {
+		toBlock = 0
+	}
+	reqs, err := f.coordinator.Requests(ctx, uint64(fromBlock), uint64(toBlock))
+	if err != nil {
+		f.logger.Errorw("Failed to fetch VRF requests",
+			"error", err,
+			"latestBlock", latestBlock,
+			"fromBlock", fromBlock,
+			"toBlock", toBlock)
+		return
+	}
+	for _, req := range reqs {
+		if _, ok := blockToRequests[req.Block]; !ok {
+			blockToRequests[req.Block] = make(map[string]struct{})
+		}
+		blockToRequests[req.Block][req.ID] = struct{}{}
+		requestIDToBlock[req.ID] = req.Block
+	}
+
+	fuls, err := f.coordinator.Fulfillments(ctx, uint64(fromBlock))
+	if err != nil {
+		f.logger.Errorw("Failed to fetch VRF fulfillments",
+			"error", err,
+			"latestBlock", latestBlock,
+			"fromBlock", fromBlock,
+			"toBlock", toBlock)
+		return
+	}
+	for _, ful := range fuls {
+		requestBlock, ok := requestIDToBlock[ful.ID]
+		if !ok {
+			continue
+		}
+		delete(blockToRequests[requestBlock], ful.ID)
+	}
+
+	for block, unfulfilledReqs := range blockToRequests {
+		if len(unfulfilledReqs) > 0 {
+			if _, ok := f.stored[block]; ok {
+				// Already stored
+				continue
+			}
+			stored, err := f.bhs.IsStored(ctx, block)
+			if err != nil {
+				f.logger.Errorw("Failed to check if block is already stored, attempting to store anyway",
+					"error", err,
+					"block", block)
+			} else if stored {
+				f.logger.Infow("Blockhash already stored",
+					"block", block, "latestBlock", latestBlock)
+				f.stored[block] = struct{}{}
+				continue
+			}
+
+			// Block needs to be stored
+			err = f.bhs.Store(ctx, block)
+			if err != nil {
+				f.logger.Errorw("Failed to store block", "error", err, "block", block)
+				continue
+			}
+			f.logger.Infow("Stored blockhash",
+				"block", block, "latestBlock", latestBlock)
+			f.stored[block] = struct{}{}
+		}
+	}
+
+	if f.lastRunBlock != 0 {
+		// Prune stored, anything older than fromBlock can be discarded
+		for block := f.lastRunBlock - uint64(f.lookbackBlocks); block < uint64(fromBlock); block++ {
+			delete(f.stored, block)
+		}
+	}
+	f.lastRunBlock = latestBlock
+}
```

Comment:
Wondering how useful it would be to have a summary log here - e.g "Stored 50 blockhashes corresponding to 60 requests" or something like that, alongside all the blocknumbers and request ID's. This way we can search a block number in LogDNA and if we see this log, that means the feeder processed it.

Reply:
For block number you could already search that and find the "Stored blockhash" log right? I think request IDs is trickier since we could be dealing with thousands, might not be practical to log all of them